utils/diagnostics.py

```diff
@@ -0,0 +1,98 @@
+
+import firedrake as fd
+
+
+def convective_cfl_calculator(mesh,
+                              name="CFL"):
+    '''
+    Return a function that calculates the cfl field from a velocity field and a timestep
+
+    :arg mesh: the mesh over which the velocity field is defined
+    :arg name: name of the CFL Function
+    '''
+    DG0 = fd.FunctionSpace(mesh, "DG", 0)
+    v = fd.TestFunction(DG0)
+
+    cell_volume = fd.Function(DG0, name="Cell volume")
+    cell_flux = fd.Function(DG0, name="Cell surface flux")
+    cfl = fd.Function(DG0, name=name)
+
+    # mesh volume
+    One = fd.Function(DG0, name="One").assign(1)
+    fd.assemble(One*v*fd.dx, tensor=cell_volume)
+
+    def both(u):
+        return 2*fd.avg(u)
+
+    def cfl_calc(u, dt):
+        # area weighted convective flux
+        n = fd.FacetNormal(u.function_space().mesh())
+        un = 0.5*(fd.inner(-u, n) + abs(fd.inner(-u, n)))
+        cell_flux_form = (
+            both(un*v)*fd.dS
+            + un*v*fd.ds
```

Comment:
need to handle the extruded mesh case
https://www.firedrakeproject.org/extruded-meshes.html#id19